mongo-store.js

```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
     },
 
     save: function (args, cb) {
-      var ent = args.ent
+      const is_update = Boolean(args.ent.id)
 
-      var update = !!ent.id
+      if (is_update) {
+        return updateExisting(args, cb)
+      }
 
-      getcoll(args, ent, function (err, coll) {
-        if (error(args, err, cb)) return
+      return createAndSave(args, cb)
 
-        var entp = ent.data$(false)
 
-        if (!update) {
-          var id
-          if (undefined !== ent.id$) {
-            id = ent.id$
-          } else if (opts.generate_id) {
-            id = opts.generate_id(ent)
+      function createAndSave(args, cb) {
+        return upsertIfRequested(args, function (err, upsert) {
+          if (error(args, err, cb)) {
+            return
           }
 
-          entp._id = makeid(id)
+          if (upsert.upsert_requested) {
+            return cb(null, upsert.out)
+          }
+
+          return createNew(args, cb)
+        })
+
+
+        function upsertIfRequested(args, cb) {
+          const query_for_save = args.q
+
+          if (Array.isArray(query_for_save.upsert$)) {
+            const upsert_on = query_for_save.upsert$
+            const public_entdata = args.ent.data$(false)
+
+            if (upsert_on.length > 0 && upsert_on.every(p => p in public_entdata)) {
+              const filter_by = upsert_on
+                .reduce((acc, field) => {
+                  acc[field] = args.ent[field]
+                  return acc
+                }, {})
+
+              const replacement = (() => {
+                const o = Dot.flatten(Object.assign({}, public_entdata))
```

Comment:
what is Dot? what does flatten do?

Reply:
@rjrodger 
Dot.flatten basically converts your object to Mongo update directives, which can be then passed to Mongo's `update*` function. E.g.:
```javascript
Dot.flatten({ foo: 'bar' }) // => { $set: { foo: 'bar' } }
```

Source: https://www.npmjs.com/package/mongo-dot-notation